packages/provider-utils/src/parse-json.test.ts

```diff
@@ -0,0 +1,190 @@
+import { describe, it, expect } from 'vitest';
+import { parseJSON, safeParseJSON, isParsableJson } from './parse-json';
+import { z } from 'zod';
+import { JSONParseError, TypeValidationError } from '@ai-sdk/provider';
+
+describe('parse-json', () => {
+  describe('parseJSON', () => {
+    it('should parse basic JSON without schema', () => {
+      const result = parseJSON({ text: '{"foo": "bar"}' });
+      expect(result).toEqual({ foo: 'bar' });
+    });
+
+    it('should parse JSON with schema validation', () => {
+      const schema = z.object({ foo: z.string() });
+      const result = parseJSON({ text: '{"foo": "bar"}', schema });
+      expect(result).toEqual({ foo: 'bar' });
+    });
+
+    it('should throw JSONParseError for invalid JSON', () => {
+      expect(() => parseJSON({ text: 'invalid json' })).toThrow(JSONParseError);
+    });
+
+    it('should throw TypeValidationError for schema validation failures', () => {
+      const schema = z.object({ foo: z.number() });
+      expect(() => parseJSON({ text: '{"foo": "bar"}', schema })).toThrow(
+        TypeValidationError,
+      );
+    });
+  });
+
+  describe('safeParseJSON', () => {
+    it('should safely parse basic JSON without schema and include rawValue', () => {
+      const result = safeParseJSON({ text: '{"foo": "bar"}' });
+      expect(result).toEqual({
+        success: true,
+        value: { foo: 'bar' },
+        rawValue: { foo: 'bar' },
+      });
+    });
+
+    it('should preserve rawValue even after schema transformation', () => {
+      const schema = z.object({
+        count: z.coerce.number(),
+      });
+      const result = safeParseJSON({
+        text: '{"count": "42"}',
+        schema,
+      });
+
+      expect(result).toEqual({
+        success: true,
+        value: { count: 42 },
+        rawValue: { count: '42' },
+      });
+    });
+
+    it('should handle failed parsing with error details', () => {
+      const result = safeParseJSON({ text: 'invalid json' });
+      expect(result).toEqual({
+        success: false,
+        error: expect.any(JSONParseError),
+      });
+    });
+
+    it('should handle schema validation failures', () => {
+      const schema = z.object({ age: z.number() });
+      const result = safeParseJSON({
+        text: '{"age": "twenty"}',
+        schema,
+      });
+
+      expect(result).toEqual({
+        success: false,
+        error: expect.any(TypeValidationError),
+      });
+    });
+
+    it('should handle nested objects and preserve raw values', () => {
+      const schema = z.object({
+        user: z.object({
+          id: z.string().transform(val => parseInt(val, 10)),
+          name: z.string(),
+        }),
+      });
+
+      const result = safeParseJSON({
+        text: '{"user": {"id": "123", "name": "John"}}',
+        schema: schema as any,
```

Comment:
why is `any` needed?

Reply:
Something about the `transform` call typing produces:

```
id: z.string().transform(val => parseInt(val, 10)), // => leads to "Type 'string' not assignable to type 'number'" error
```

for this test case `z.coerce.number()` achieves the same thing and doesn't produce the type issue, so I'll change it to that (we did similar in another test case).

Reply:
that's a bug and it has to do with the zod schemas

Reply:
I think I know what it is; can take a look next week (zod has input and output schemas)

Reply:
let's leave any for now